File: xiaoming/0917/reverseLinkedList.java

public class reverseLinkedList {
	
	public static void main(String[] args) {
		ListNode l1 = new ListNode(20);
		l1.next = new ListNode(60);
		
		ListNode l2 = reverseList(l1);
		
		System.out.println(l2.val+ "  "+l2.next.val+ "  ");
	}
	
	public static ListNode reverseList(ListNode head) {
		
	    if(head == null)  return null;
	    if(head.next == null) return head;
	    
	    ListNode helper = new ListNode(0);
	    helper.next = head;
	    ListNode pre = helper;
	    ListNode cur = head.next;
	    while(cur!=null)
	    {
	        ListNode next = cur.next;
	        cur.next = pre.next;
	        pre.next = cur;
	        head.next = next;
	        cur = next;
	    }
	    return helper.next;
	}
}

//class ListNode {
//    int val;
//    ListNode next;
//    ListNode() { 
//    	val = 0;
//    	next = null;
//    }
//    ListNode(int x) { 
//    	val = x;
//    	next = null;
//    }
// }

File: mu_wang/9_17/Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        new_head = ListNode(0)
        if not head:
            return 
        flag = True
        temp = list()
        while head:
            temp.append(ListNode(head.val))
            head = head.next
            if flag:
                flag = False
                continue
            else:
                temp[-1].next = temp[-2]
        new_head = temp[-1]
        return new_head

    def reverseList_2(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        new_head = ListNode(0)
        if not (head and head.next):
            return head
        rst =None
        pre = None
        next_ = None
        while head :
            if rst is None:
                rst = head
                next_ = head.next
                rst.next = pre
                pre = rst
            else:
                rst = head
                next_ = head.next
                rst.next = pre
                pre = rst
            head = next_
        return rst
